src/blop/agent.py

```diff
@@ -542,32 +546,86 @@ def posterior(self, x):
+            if len(self.active_objs) > 1
+            else self.active_objs[0].model
         )
 
     @property
-    def scalarized_objectives(self):
-        """Returns a (n_obs,) array of scalarized objectives"""
-        return self.targeting_transform.evaluate(self.train_targets(active=True)).sum(axis=-1)
+    def fitness_weights(self):
+        return torch.tensor([obj.weight for obj in self.objectives if obj.type == "fitness"], dtype=torch.double)
 
     @property
-    def max_scalarized_objective(self):
-        """Returns the value of the best scalarized objective seen so far."""
-        f = self.scalarized_objectives
-        return np.max(np.where(np.isnan(f), -np.inf, f))
+    def fitness_scalarization(self):
+        return ScalarizedPosteriorTransform(weights=self.fitness_weights)
 
     @property
-    def argmax_scalarized_objective(self):
-        """Returns the index of the best scalarized objective seen so far."""
-        f = self.scalarized_objectives
-        return np.argmax(np.where(np.isnan(f), -np.inf, f))
+    def evaluated_fitnesses(self):
+        return self.train_targets()[:, self.objectives.type == "fitness"]
+
+    @property
+    def scalarized_fitnesses(self):
+        return self.fitness_scalarization.evaluate(self.evaluated_fitnesses)
+
+    @property
+    def evaluated_constraints(self):
+        if sum(self.objectives.type == "constraint"):
+            y = self.train_targets()
+            return torch.cat(
+                [
+                    ((y[:, i] >= obj.target[0]) & (y[:, i] <= obj.target[1])).unsqueeze(0)
+                    for i, obj in enumerate(self.objectives)
+                    if obj.type == "constraint"
+                ],
+                dim=0,
+            ).T
+        else:
+            return torch.ones(size=(len(self.table), 0), dtype=torch.bool)
+
+    @property
+    def argmax_best_f(self):
+        f = self.scalarized_fitnesses
+        c = self.evaluated_constraints.all(axis=-1)
+        mask = (~f.isnan()) & c
+
+        if not mask.sum():
+            raise ValueError("There are no valid points that satisfy the constraints!")
+
+        return torch.where(mask)[0][f[mask].argmax()]
+
+    @property
+    def best_f(self):
+        return self.scalarized_fitnesses[self.argmax_best_f]
+
+    @property
+    def pareto_front_mask(self):
+        # a point is on the Pareto front if it is Pareto dominant
+        # a point is Pareto dominant if it is there is no other point that is better at every objective
+        y = self.train_targets()[:, self.objectives.type == "fitness"]
+        in_pareto_front = ~(y.unsqueeze(1) > y.unsqueeze(0)).all(axis=-1).any(axis=0)
+        all_constraints_satisfied = self.evaluated_constraints.all(axis=-1)
+        return in_pareto_front & all_constraints_satisfied
+
+    @property
+    def pareto_front(self):
+        return self.table.loc[self.pareto_front_mask.numpy()]
```

Comment:
I may need some explanation on the `pareto_front_mask` and `pareto_front`. Maybe a docstring can have that explanation?